### PR TITLE
Improve message composition

### PR DESCRIPTION
**Pull request by szepeviktor**
Use sprintf instead of complex concatenations.

### CODE REVIEW

File: src/Loader/Http/HttpLoader.php [hunk 1]
```diff
@@ -101,16 +101,22 @@ public function __construct(
...
+                return;
             }
 
-            $logger->error($logMessage);
+            $logger->error(sprintf(
+                'Failed to load %s: got response %s - %s',
+                $request->getUri()->__toString(),
+                $exceptionOrResponse->getStatusCode(),
+                $exceptionOrResponse->getReasonPhrase()
+            ));
```
Comment (szepeviktor) [inline src/Loader/Http/HttpLoader.php:119]: This is straightforward, the whole message in one place.

File: src/Loader/Http/Politeness/RetryErrorResponseHandler.php [hunk 2]
```diff
@@ -138,14 +138,16 @@ protected function logRepeatedErrorMessage(RespondedRequest $respondedRequest):
 
     /**
      * @throws LoadingException
+     * @return never
      */
-    protected function retryAfterExceedsLimitMessage(ResponseInterface $response): string
+    protected function retryAfterExceedsLimitMessage(ResponseInterface $response)
```
Comment (szepeviktor) [inline src/Loader/Http/Politeness/RetryErrorResponseHandler.php:143]: This method always throws an exception, never returns a value.

File: src/Loader/Http/Politeness/RetryErrorResponseHandler.php [hunk 3]
```diff
@@ -138,14 +138,16 @@ protected function logRepeatedErrorMessage(RespondedRequest $respondedRequest):
...
     {
         $statusCodeAndReasonPhrase = $this->getResponseCodeAndReasonPhrase($response);
 
-        $message = 'Retry-After header in ' . $statusCodeAndReasonPhrase . ' response, requires to wait longer ' .
-            'than the defined max wait time for this case. If you want to increase this limit, set it ' .
-            'in the ErrorResponseHandler of your HttpLoader instance.';
+        $message = sprintf(
+            'Retry-After header in %s response indicates a too long wait, see your HttpLoader::retryErrorResponseHandler',
```
Comment (szepeviktor) [inline src/Loader/Http/Politeness/RetryErrorResponseHandler.php:148]: Improved message with exception-like lenght.

File: src/Stores/JsonFileStore.php [hunk 4]
```diff
@@ -38,7 +38,11 @@ public function store(Result $result): void
...
     public function filePath(): string
     {
-        return $this->storePath . '/' .
-          ($this->filePrefix ? $this->filePrefix . '-' : '') . $this->createTimestamp . '.json';
+        return sprintf(
+            '%s/%s%s.json',
+            $this->storePath
+            $this->filePrefix ? $this->filePrefix . '-' : '',
+            $this->createTimestamp
+        );
```
Comment (szepeviktor) [inline src/Stores/JsonFileStore.php:46]: No concatenation operator, you can see all three parts separately.

File: src/Loader/Http/HttpLoader.php (hunk 1 above)
Comment (szepeviktor) [reply src/Loader/Http/HttpLoader.php:119]:
```
if (error) unhappy path;

happy path;
```

[_source_](https://github.com/szepeviktor/very-clean-code)

File: src/Stores/JsonFileStore.php [hunk 5]
```diff
@@ -38,7 +38,11 @@ public function store(Result $result): void
 
     public function filePath(): string
     {
-        return $this->storePath . '/' .
-          ($this->filePrefix ? $this->filePrefix . '-' : '') . $this->createTimestamp . '.json';
+        return sprintf(
+            '%s/%s%s.json',
```
Comment (szepeviktor) [inline src/Stores/JsonFileStore.php:42]: The last placeholder could actually be `%d` as `createTimestamp` is an integer.

**Comment by otsch** [comment]
I know it's a bad preference of mine to do string concatenation like this. But actually for me it still feels better than `sprintf`, because when reading, the inserted variables are right in the place where they are inserted and I don't have to count and see which argument will be inserted for the third %s.

```php
$number = 123;

$string = 'asdf';

$message = 'Foo ' . $number . ' bar ' . $string;

$message = sprintf('Foo %s bar %s', $number, $string);

$message = "Foo {$number} bar {$string}";
```

I'll accept it if you go for the third option.

**Comment by szepeviktor** [comment]
> still feels better

When I meet feeling 🐰 I bravely run away

# 🏃🏻‍♂️

**Comment by otsch** [comment]
@szepeviktor OK, I used the word "feels", but actually I also argued why I prefer concatenation over `sprintf`. It's not just a preference without any reason and if it was, still no reason to be rude.

**Comment by szepeviktor** [comment]
@otsch I see. When I meet personal things, preferences I **really** run away.
Viktor is 100% data-driven even in things not to my liking.

`printf` was invented for this purpose. Everyone has feelings and preferences.
I put them away when it comes to software making.

**Comment by szepeviktor** [comment]
ps. In the pas decade I have learns liking measurement results. Some of them are still generating bad feelings!

**Comment by szepeviktor** [comment]
> still no reason to be rude.

I am very sorry that you felt rudeness.
I just wanted to let you know that I do not talk about personal things. I only mention facts, measurement results.

**Comment by otsch** [comment]
OK, so what could you measure to decide which option is better here? Isn't it mainly about readability?
What makes `sprintf` the superior choice for you over the mentioned third option? 

> ```php
> $number = 123;
> 
> $string = 'asdf';
> 
> $message = 'Foo ' . $number . ' bar ' . $string;
> 
> $message = sprintf('Foo %s bar %s', $number, $string);
> 
> $message = "Foo {$number} bar {$string}";
> ```

**Comment by szepeviktor** [comment]
Very good question. Thank you!

Every language has bad parts. When some parts are targeting developer experience, not the robustness of the code.

The concatenation operator is [a bad choice](https://github.com/szepeviktor/phpcs-psr-12-neutron-hybrid-ruleset/blob/cc03e7cec95feab69f4ad9a1124b415911d5d1b4/PSR12NeutronRuleset/Sniffs/Strings/ConcatenationUsageSniff.php#L42). It mixes literal strings and values of variables. It is the typical "too much liberty".

The _complex curly syntax_ (as PHP docs call it) is the worst. It mixes string literals and variable values **deeper**. There is barely any separation between them. And curly braces are highly error prone: just a single space and it stops working, no QA tool will notice that unless you unit-test all parts of it.

Then it comes to good old `printf`.
1. It 100% separates string literals and variable values.
2. QA tools can check number of placeholders, type of placeholders vs. type of variables
3. You can put every variable on a separate line: 1 line 1 variable

Side effects
- it takes a year or two to read (to be familiar with) printf syntax like plain text
- feeling bad may be more important than making a software robust
- it takes a bit more time to write printf

ps. PHP was born as a template engine, and since 7.0 it wants to be a Java-like programming language.

**Comment by szepeviktor** [comment]
Finally something personal. Secret! I do not navigate by my personal things.
```php
$message =
    'Foo '
    . $number
    . ' bar '
    . $string;
```
This is to my linking ❤️ 
But you won't see me doing it.

**Comment by otsch** [comment]
Sorry, but I think there's a lot of preference and believes in your argumentation:
* The "separation" that you're speaking of, isn't that a pretty abstract thing? In the end, all three methods insert strings (or integers, or anything that can somehow be casted to a string) into another string. With sprintf there's just a placeholder inside the string literal where it is inserted, instead of the variable itself.
* "It can handle many types not just strings" => 🤷🏻‍♂️ you can also use integers or other things in string concatenation. Haven't experienced any problems with that. And also I'm not really sure about the types in sprintf...it's basically only forced type casting. But what's the value of that if I expect some variable to be a number, but the actual incoming variable is a string like `abc`, then `sprintf('Foo %d', $var)` will result in `Foo 0`. I think handling types is better done with type hints and return types.
* `no QA tool will notice that unless you unit-test all parts of it.` => in your comment you linked to a custom PHP CS fixer rule that you've built 😅 so, if you know that you sometimes make mistakes using the curly braces syntax, why not just make a CS fixer rule for that?

And further you proofed yourself that you can put everything on a separate line also with string concatenation...which I think is just a matter of taste not an argument for printf superiority. And also you mentioned yourself that it takes a year or two, to read printf syntax like plaintext. I think I just did this with string concatenation, which basically is a pain for almost everyone, but I'm just used to it. But I admit, that the curly braces syntax definitely looks better.

**Pull request closed without merging**